node/pkg/driver/node.go

```diff
@@ -613,9 +626,67 @@ func (d *NodeService) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
 }
 
 func (d *NodeService) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	goid_info.SetAdditionalIDInfo(req.VolumeId)
+	volumeId := req.VolumeId
+	goid_info.SetAdditionalIDInfo(volumeId)
 	defer goid_info.DeleteAdditionalIDInfo()
-	return nil, status.Error(codes.Unimplemented, "NodeGetVolumeStats is not implemented yet")
+	volumePath := req.VolumePath
+	volumePathWithHostPrefix := d.NodeUtils.GetPodPath(req.VolumePath)
+	isFSVolume := true
+
+	err := d.nodeGetVolumeStatsRequestValidation(volumeId, volumePath)
+	if err != nil {
+		return nil, err
+	}
+
+	isPathExists := d.NodeUtils.IsPathExists(volumePathWithHostPrefix)
+	if !isPathExists {
+		return nil, status.Errorf(codes.NotFound, "volume path %q does not exists", volumePath)
+	}
+	isBlock, err := d.NodeUtils.IsBlock(volumePathWithHostPrefix)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to determine if %q is block device: %s", volumePath, err)
+	}
+	if isBlock {
+		isFSVolume = false
+	}
+	isMounted, err := d.IsMounted(volumePathWithHostPrefix, isFSVolume)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to check if volume path %q is mounted", volumePath)
+	}
+	if !isMounted {
+		return nil, status.Errorf(codes.NotFound, "volume path %q is not mounted", volumePath)
+	}
+	volumeStats, err := d.NodeUtils.GetVolumeStats(volumePathWithHostPrefix, isFSVolume)
```

Comment:
according to the [spec](https://github.com/container-storage-interface/spec/blob/v1.5.0/spec.md#nodegetvolumestats), regarding the `VolumePath`:
>  // It can be any valid path where volume was previously
  // staged or published.

and we currently [don't](https://github.com/IBM/ibm-block-csi-driver/blob/v1.8.0/node/pkg/driver/node.go#L159) mount block volumes on stage, so I think it would be better to find the relevant path from the volume id (we need it for [NOT_FOUND](https://github.com/IBM/ibm-block-csi-driver/pull/457#discussion_r810976948) [grpc code](https://github.com/container-storage-interface/spec/blob/v1.5.0/spec.md#nodegetvolumestats-errors) anyway)

Reply:
how can I get the volume path from the volume id? it doesn't have a path in the volume id?

Reply:
not sure, but there is `GetMpathDevice`, and from there, there should be a way

Reply:
opened a ticket for it as we discussed CSI-3946